pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -928,6 +943,27 @@ public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
         }
     }
 
+    private int removeExpiredMessagesFromQueue(Set<MessageIdImpl> messageIds) {
+        int messagesFromQueue = 0;
+        for (Message message : incomingMessages) {
```

Comment:
I think you already mentioned somewhere that iterating over the custom `GrowableArrayBlockingQueue` is not currently working. We need the queue to be growable to supported batch messages, since the broker doesn't know the exact messages count when it reads a single entry from BK.